text/3185-static-async-fn-in-trait.md

```diff
@@ -0,0 +1,258 @@
+}
+```
+
+## Semantic rules
+
+When an async function is present in a trait or trait impl...
+
+### The trait is not considered dyn safe
+
+This limitation is expected to be lifted in future RFCs.
+
+### Both the trait and its impls must use `async` syntax
+
+It is not legal to use an async function in a trait and a "desugared" function in an impl.
+
+## Equivalent desugaring
+
+### Trait
+
+Async functions in a trait desugar to an associated function that returns a generic associated type (GAT):
+
+* Just as with [ordinary async functions](https://rust-lang.github.io/rfcs/2394-async_await.html#lifetime-capture-in-the-anonymous-future), the GAT has a generic parameter for every generic parameter that appears on the fn, along with implicit lifetime parameters.
+* The GAT has the complete set of where clauses that appear on the `fn`, including any implied bounds.
+* The GAT is "anonymous", meaning that its name is an internal symbol that cannot be referred to directly. (In the examples, we will use `$` to represent this name.)
+
+
+```rust
+trait Example {
+    async fn method<P0..Pn>(&self)
+    where
+        WC0..WCn;
+}
+
+// Becomes:
+
+trait Example {
+    type $<'me, P0..Pn>: Future<Output = ()>
+    where
+        WC0..WCn, // Explicit where clauses
+        Self: 'me; // Implied bound from `&self` parameter
+
+    fn method<P0..Pn>(&self) -> Self::$<'_, P0..Pn>
+    where
+        WC0..WCn;
+}
+```
+
+`async fn` that appear in impls are desugared in the same general way as an [existing async function](https://doc.rust-lang.org/reference/items/functions.html#async-functions), but with some slight differences:
+
+* The value of the associated type `$` is equal to an `impl Future` type, rather than the `impl Future` being the return type of the function
+* The function returns `Self::$<...>` with all the appropriate generic parameters
+
+Otherwise, the desugaring is the same. The body of the function becomes an `async move { ... }` block that both (a) captures all parameters and (b) contains the body expression.
+
+```rust
+impl Example for ExampleType {
+    async fn method<P0..Pn>(&self) {
+        ...
+    }
+}
+
+impl Example for ExampleType {
+    type $<'me, P0..Pn> = impl Future<Output = ()> + 'me
+    where
+        WC0..WCn, // Explicit where clauses
+        Self: 'me; // Implied bound from `&self` parameter
+
+    fn method<P0..Pn>(&self) -> Self::$<'_, P0..Pn> {
+        async move { ... }
+    }
+}
+```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Why are we adding this RFC now?
+
+This RFC represents the least controversial addition to async/await that we could add right now. It was not added before due to limitations in the compiler that have now been lifted – namely, support for [Generic Associated Types][gat] and [Type Alias Impl Trait][tait].
+
+[gat]: https://github.com/rust-lang/generic-associated-types-initiative
+[tait]: https://github.com/rust-lang/rust/issues/63063
```

Comment:
[tait] should [trait]

Reply:
No, `tait` (usually uppercase) is the abbreviation for type alias impl trait.